File: app/src/main/java/com/neubofy/veto/ui/MainActivity.java
package com.neubofy.veto.ui;

import static com.neubofy.veto.ui.SetupWarningsActivityKt.shouldShowSetupWarnings;
import static com.neubofy.veto.ui.UiUtil.setupEdgeToEdgeAppBar;

import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;

import androidx.annotation.NonNull;

import com.google.android.material.appbar.MaterialToolbar;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;

import com.neubofy.veto.BuildConfig;
import com.neubofy.veto.R;
import com.neubofy.veto.data.Settings;
import com.neubofy.veto.data.SettingsRepository;
import com.neubofy.veto.services.TempContactExpiredService;
import com.neubofy.veto.ui.home.MainPageFragment;

import com.neubofy.veto.ui.settings.AboutActivity;
import com.neubofy.veto.utils.UpdateManager;
import kotlin.Unit;

public class MainActivity extends VetoActivity {

    SettingsRepository settings;

    public static final String EXTRA_OPEN_FRAGMENT = "EXTRA_OPEN_FRAGMENT";

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        setIntent(intent);
        handleIntent(intent);
    }

    private void handleIntent(Intent intent) {
        if (intent != null && "PERMISSIONS".equals(intent.getStringExtra(EXTRA_OPEN_FRAGMENT))) {
            int highlightName = intent.getIntExtra(com.neubofy.veto.ui.home.PermissionManagerFragment.ARG_HIGHLIGHT_PERMISSION_NAME, -1);
            com.neubofy.veto.ui.home.PermissionManagerFragment fragment = new com.neubofy.veto.ui.home.PermissionManagerFragment();
            if (highlightName != -1) {
                Bundle args = new Bundle();
                args.putInt(com.neubofy.veto.ui.home.PermissionManagerFragment.ARG_HIGHLIGHT_PERMISSION_NAME, highlightName);
                fragment.setArguments(args);
            }
            getSupportFragmentManager().beginTransaction()
                    .replace(R.id.fragment_container, fragment)
                    .addToBackStack(null)
                    .commit();
        }
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_main);

        MaterialToolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        setupEdgeToEdgeAppBar(findViewById(R.id.appBar));

        settings = SettingsRepository.Companion.getInstance(this);
        settings.load();

        if (((Integer) settings.get(Settings.SET_APP_CRASHED_LOG_ENTRY)) == 1) {
            settings.set(Settings.SET_APP_CRASHED_LOG_ENTRY, 0);
            com.neubofy.veto.utils.Notifications.notify(
                this,
                "Veto Background Recovery",
                "App recovered from an uncaught background crash. Log entry saved in Log View.",
                com.neubofy.veto.utils.Notifications.CHANNEL_FAILED,
                com.neubofy.veto.ui.settings.LogViewActivity.class
            );
        }

        if (savedInstanceState == null) {
            getSupportFragmentManager().beginTransaction()
                    .replace(R.id.fragment_container, new MainPageFragment())
                    .commit();
            handleIntent(getIntent());
        }

        // Silently check for OTA updates
        UpdateManager.INSTANCE.checkForUpdates(this, true, false, null);
    }

    @Override
    protected void onResume() {
        super.onResume();
        TempContactExpiredService.scheduleJob(this, 0);
    }
}

File: app/src/main/java/com/neubofy/veto/ui/MainActivity.kt
package com.neubofy.veto.ui

import android.content.Intent
import android.os.Bundle
import androidx.activity.compose.setContent
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.ui.Modifier
import androidx.compose.runtime.remember
import androidx.navigation.compose.NavHost
import androidx.navigation.compose.composable
import androidx.navigation.compose.rememberNavController
import com.neubofy.veto.R
import com.neubofy.veto.data.Settings
import com.neubofy.veto.data.SettingsRepository
import com.neubofy.veto.services.TempContactExpiredService
import com.neubofy.veto.ui.home.MainPageScreen
import com.neubofy.veto.ui.home.TransportListScreen
import com.neubofy.veto.ui.home.PermissionManagerScreen
import com.neubofy.veto.ui.settings.LogViewActivity
import com.neubofy.veto.ui.theme.VetoTheme
import com.neubofy.veto.utils.Notifications
import com.neubofy.veto.utils.UpdateManager
import com.neubofy.veto.transports.availableTransports
import com.neubofy.veto.permissions.globalAppPermissions

class MainActivity : VetoActivity() {

    private lateinit var settings: SettingsRepository

    companion object {
        const val EXTRA_OPEN_FRAGMENT = "EXTRA_OPEN_FRAGMENT"
    }

    // A simple hack to expose the required highlighted permission for deep linking
    private var initialPermissionHighlightName: Int = -1
    private var isNavigationPending = androidx.compose.runtime.mutableStateOf(false)
    private var pendingDestination = "main"

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        setIntent(intent)
        handleIntent(intent)
    }

    private fun handleIntent(intent: Intent?) {
        if (intent != null && "PERMISSIONS" == intent.getStringExtra(EXTRA_OPEN_FRAGMENT)) {
            // ARG_HIGHLIGHT_PERMISSION_NAME equivalent handling
            // The legacy system used the string literal "ARG_HIGHLIGHT_PERMISSION_NAME"
            initialPermissionHighlightName = intent.getIntExtra("ARG_HIGHLIGHT_PERMISSION_NAME", -1)
            pendingDestination = "permissions"
            isNavigationPending.value = true
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        settings = SettingsRepository.getInstance(this)
        settings.load()

        if (settings.get(Settings.SET_APP_CRASHED_LOG_ENTRY) == 1) {
            settings.set(Settings.SET_APP_CRASHED_LOG_ENTRY, 0)
            Notifications.notify(
                this,
                "Veto Background Recovery",
                "App recovered from an uncaught background crash. Log entry saved in Log View.",
                Notifications.CHANNEL_FAILED,
                LogViewActivity::class.java
            )
        }

        handleIntent(intent)

        setContent {
            VetoTheme {
                Surface(
                    modifier = Modifier.fillMaxSize(),
                    color = MaterialTheme.colorScheme.background
                ) {
                    val navController = rememberNavController()

                    // Handle intent deep linking manually on launch
                    androidx.compose.runtime.LaunchedEffect(isNavigationPending.value) {
                        if (isNavigationPending.value) {
                            navController.navigate(pendingDestination) {
                                popUpTo("main") { inclusive = false }
                            }
                            isNavigationPending.value = false
                        }
                    }

                    NavHost(navController = navController, startDestination = "main") {
                        composable("main") {
                            MainPageScreen(
                                onOpenCommands = {
                                    startActivity(Intent(this@MainActivity, com.neubofy.veto.ui.settings.CommandsActivity::class.java))
                                },
                                onOpenSettings = {
                                    startActivity(Intent(this@MainActivity, com.neubofy.veto.ui.settings.SettingsActivity::class.java))
                                },
                                onOpenPermissions = {
                                    navController.navigate("permissions")
                                },
                                onOpenTransports = {
                                    navController.navigate("transports")
                                }
                            )
                        }

                        composable("transports") {
                            TransportListScreen(
                                transports = availableTransports(this@MainActivity),
                                activity = this@MainActivity
                            )
                        }

                        composable("permissions") {
                            PermissionManagerScreen(
                                permissions = globalAppPermissions(),
                                activity = this@MainActivity,
                                highlightName = initialPermissionHighlightName
                            )
                        }
                    }
                }
            }
        }

        // Silently check for updates
        UpdateManager.checkForUpdates(this, silent = true, isBeta = false, onCheckComplete = null)
    }

    override fun onResume() {
        super.onResume()
        TempContactExpiredService.scheduleJob(this, 0)
    }
}
